### PR TITLE
rose suite-run: create share/cycle/ sub-directory

### LINKED ISSUE #1333: Specify location for cycling files
It would be useful if we could separate out the share directory into two areas.
 1. An area for files which are shared only between cycles, which are regularly house-kept.
 2. An area for files which are shared throughout the whole suite.

This would allow different storage areas to be used. e.g. 1, could be directed to a storage area which is deleted more aggressively than 2.

### PR DESCRIPTION
**Pull request by matthewrmshin**
This sub-directory can be set to make use of a less permanent filesystem.

Close #1333.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
Under the proposed change, `ROSE_DATAC` (and friends) will point to cycle point directories under `${HOME}/cylc-run/${SUITE}/sharecycle/`. Current behaviour can be mimicked by adding a configuration in site/user `rose.conf`:

```
[rose-suite-run]
root-dir-sharecycle=*=share/data
```

which will make `sharecycle/` a symbolic link of `share/data/`.

**Comment by matthewrmshin** [comment]
@arjclark please review and test.

File: t/rose-task-run/06-app-prune-iso.t [hunk 1]
```diff
@@ -57,6 +57,7 @@ fi
 TEST_KEY=$TEST_KEY_BASE-prune-log
 sed '/^\[INFO\] \(create\|delete\|update\)/!d;
      /^\[INFO\] create.*share\/data/d;
```
Comment (arjclark) [inline t/rose-task-run/06-app-prune-iso.t]: Still necessary given removal of share/data?

**Comment by arjclark** [comment]
@benfitzpatrick - I'm nominally happy with the code changes. Given this is effectively a major change, can you cast a second pair of eyes over this to ensure it's sound?

**Comment by arjclark** [comment]
@matthewrmshin - am consistently getting test failures with these tests:

```
t/rose-suite-hook/03-mail.t       (Wstat: 0 Tests: 31 Failed: 3)
  Failed tests:  8, 17, 26
t/rose-task-run/05-app-prune.t    (Wstat: 0 Tests: 9 Failed: 4)
  Failed tests:  3, 7-9
t/rose-task-run/06-app-prune-iso.t (Wstat: 0 Tests: 7 Failed: 4)
  Failed tests:  1, 5-7
t/rose-suite-clean/03-only-2.t    (Wstat: 0 Tests: 4 Failed: 2)
  Failed tests:  2, 4
```

**Comment by matthewrmshin** [comment]
For these ones:

```
t/rose-task-run/05-app-prune.t
t/rose-task-run/06-app-prune-iso.t
t/rose-suite-clean/03-only-2.t
```

You need to ensure that any remote host is using this version of Rose.

I cannot see why `t/rose-suite-hook/03-mail.t` is failing. This change should not affect it.

**Comment by arjclark** [comment]
@matthewrmshin - sorted. Looks like the problem was with my hpc copy not being lined up correctly.

Still can't get the mail test to pass on your branch but it is at least failing for me on master too now.

**Comment by arjclark** [comment]
@benfitzpatrick - can you test and feedback too now?

**Comment by arjclark** [comment]
When it comes to writing the Changes.md file this will need flagging in big bold letters, particular with a view to the (effectively) backwards compatibility setting in the site config.

**Comment by benfitzpatrick** [comment]
(The mail test is very sensitive, for reasons that are not obvious. I'm working on fixing it in a separate piece of work.)

The rest of the Rose user guide needs updating.

Otherwise looks OK.

@hjoliver, what do you think of the concept and the naming?

**Comment by matthewrmshin** [comment]
> The rest of the Rose user guide needs updating.

I have done a search for `share/` and `work/` in the docs. Where relevent, I have updated the docs with entries for `sharecycle/`.

**Comment by hjoliver** [comment]
I like the general idea, which I presume is mostly to enable Rose utilities to distinguish between cycle point and  cross-cycle point share directories?

(So long as utilities like ```rose_prune``` can also be configured to work with other user-defined locations too - see https://github.com/metomi/rose/issues/1333#issuecomment-69526380).

I'm not sure I like the name and location though - "sharecycle" is a bit ungainly, and the location makes the suite run directory a bit less tidy. Shouldn't all share spaces (particularly when you consider that users may sometimes need to use other custom locations as per my comment referenced above) appear as sub-directories of the general suite share directory?  What about ```share/cycles/``` and ```share/misc/```  (or something else for "misc")?

**Comment by matthewrmshin** [comment]
All considered, `sharecycle/` is easiest to implement without harming any existing functionality. It is a variant of `share/`, a bit like `lib/` and `libexec/` in a standard file system hierarchy.

**Comment by matthewrmshin** [comment]
#1593 will follow this with a change to `rose_prune` to allow it to prune any item under the suite run time directory.

**Comment by benfitzpatrick** [comment]
I'm not 100% on 'sharecycle' either (although I can't think of anything better).

> sharecycle/ is easiest to implement

but just to be clear, is `share/cycles` possible, impossible, or too difficult?

**Comment by matthewrmshin** [comment]
Nothing is impossible, but a root level directory is least complicated to implement and configure.

**Comment by benfitzpatrick** [comment]
I had thought that it would be too impractical.

In that case, if it is not super-horrible, I would prefer it as a subdirectory. I'd rather have it nice for the users.

**Comment by hjoliver** [comment]
Ease of implementation is but one consideration ... but of course I'm not task with doing it :grin:  

From a purely semantic perspective though, "share" should be more general than "sharecycle", but you're relegating it to, in effect, "sharenotcycle".  Also, this seems to break with the standard cylc run tree layout a bit, whereas sub-directories of share would fit it.

**Comment by matthewrmshin** [comment]
Just to put the scary bit in, the configuration and the location of the symbolic links would look horrible to the users.

**Comment by matthewrmshin** [comment]
OK, popular demand wins. I have to change the `root-dir-*` configuration syntax slightly to `root-dir{*}` to make it look less ugly. (The old syntax can still be used.) This will also match what I have done in #1593.

Backward compatibility for `ROSE_DATAC` can now be achieved by doing:

```cfg
[rose-suite-run]
root-dir{share/cycle}=*=share/data
```

**Comment by matthewrmshin** [comment]
I have squashed and rebased the branch, as previous commits make no sense any more.

**Comment by benfitzpatrick** [comment]
I'm happy with this and will soon merge it if the test battery passes.

**Pull request merged**